### PR TITLE
cpu limits for network-costs

### PR DESCRIPTION
**Pull request by jessegoodier**
Please review.

### CODE REVIEW

**Review by AjayTripathy — APPROVED** [review]

**Comment by dwbrown2** [comment]
There's a lot of discussion about cpu limits -- https://home.robusta.dev/blog/stop-using-cpu-limits

Do we want to actively recommend?

File: network-allocation.md [hunk 1]
```diff
@@ -5,16 +5,39 @@ This document summarizes Kubecost network cost allocation, how to enable it, and
...
+
+ Alternatively a serviceMonitor is also [available](https://github.com/kubecost/cost-analyzer-helm-chart/blob/700cfa306c8e78bc9a1039b584769b9a0e0757d0/cost-analyzer/values.yaml#L716).
+
 > **Note**: Network cost, which is disabled by default, needs to be run as a privileged pod to access the relevant networking kernel module on the host machine.
- 
+
+## Resource Limiting
+
+In order to reduce resource usage, Kubecost recommends setting a CPU limit on the network-costs daemonset. This will cause a few seconds delay during peak usage and does not effect overall accuracy. This is done by default in Kubecost 1.99+.
+
```
Comment (bstuder99) [inline network-allocation.md:29]: Are we holding off on merging this until 1.99 then?

**Review by bstuder99 — APPROVED** [review]

File: network-allocation.md (hunk 1 above)
Comment (bstuder99) [reply network-allocation.md:29]: If so, we should hold off on merging and tag this PR

**Comment by mbolt35** [comment]
> There's a lot of discussion about cpu limits -- https://home.robusta.dev/blog/stop-using-cpu-limits
> 
> Do we want to actively recommend?

The author does a great job explaining (both via the analogy AND the technical writeup). Although, he's looking at this from the "owner" of a cluster. In the context of running your own infrastructure and understanding how kubernetes works when requests and limits are set (or not set), then I 100% agree with this post. However, when our feedback from a customer is that "network costs is using a lot of CPU", we only have a few options. I would love to provide them information about how to restrict the network-costs CPU with requests, which will maximize their CPU usage (especially if there is excess available). I think "high CPU" gets misunderstood, like the article mentions, as "it's taking CPU away from other pods/containers." If we try to provide a corrective measure or point our users towards more information, and it isn't well received, then CPU limits does feel like the only remaining tool to guarantee that a pod doesn't exceed some threshold. 

I do feel strongly we should be guiding our users towards proper kubernetes management, but when this fails, it does seem reasonable to leverage less favorable options to ensure that their needs are met. 

@jessegoodier You may want to just provide a note referencing the content (or article) @dwbrown2 mentioned. I can confirm if the pod is hard limited, then it should function appropriately as long as it is not completely starved, so setting limits will work if the only goal is to prevent the pod from using "high CPU."

**Comment by jessegoodier** [comment]
> There's a lot of discussion about cpu limits -- https://home.robusta.dev/blog/stop-using-cpu-limits
> 
> Do we want to actively recommend?

I have read that- very interesting.

I think this case may be the exception to the article.

My thinking is that a workload that is not latency sensitive (like a cumulative metric endpoint) can be CPU limited with no impact to the value the workload provides. 

The alternative is that no-limits will cause actual workloads that are latency sensitive to compete for CPU resources exactly when the system is under load.

And from a support perspective we have customers that are noticing that this is an expensive workload- questioning the value of running it.

I'm not an expert here, so totally open to killing this change.

File: network-allocation.md (hunk 1 above)
Comment (jessegoodier) [reply network-allocation.md:29]: I'll have a draft PR for the change in the helm chart. No harm in publishing this doc now given we have instructions for implementing the change.

File: network-allocation.md [hunk 2]
```diff
@@ -5,16 +5,40 @@ This document summarizes Kubecost network cost allocation, how to enable it, and
...
- 
+ You can view a list of common config options [here](https://github.com/kubecost/cost-analyzer-helm-chart/blob/700cfa306c8e78bc9a1039b584769b9a0e0757d0/cost-analyzer/values.yaml#L573).
+
+ If using the included Prometheus instance, the scrape is automatically conifgured.
+
+ If you are integrating with an existing Prometheus, you can set `networkCosts.prometheusScrape=true` and the network costs service should be auto-discovered.
+
+ Alternatively a serviceMonitor is also [available](https://github.com/kubecost/cost-analyzer-helm-chart/blob/700cfa306c8e78bc9a1039b584769b9a0e0757d0/cost-analyzer/values.yaml#L716).
+
 > **Note**: Network cost, which is disabled by default, needs to be run as a privileged pod to access the relevant networking kernel module on the host machine.
```
Comment (jessegoodier) [inline network-allocation.md:24]: @dwbrown2 any changes that you'd recommend?

File: network-allocation.md [hunk 3]
```diff
@@ -5,16 +5,40 @@ This document summarizes Kubecost network cost allocation, how to enable it, and
...
+
+```
+    resources:
+      limits:
+        cpu: 500m
+      requests:
+        cpu: 50m
+        memory: 20Mi
+```
+ > This write-up regarding CPU limiting <https://home.robusta.dev/blog/stop-using-cpu-limits> has points worth considering. In the case of low-priority/latency insensitive jobs, there is little risk to setting CPU limits.
```
Comment (jessegoodier) [inline network-allocation.md:40]: any changes that you'd recommend?

Comment (jessegoodier) [reply network-allocation.md]: I removed the note.

I think it causes more confusion than clarity. 

I'm open to adding more color if we think it is needed.

**Pull request merged**